lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,1314 @@
+// perform actions (only send messages) via defined state transitions that do not violate the
+// protocol.
+//
+// An example of a full negotiation and associated states follows:
+//
+//     +------------+                         +------------------+---- Holder state after message sent/received ----+
+//     |            |--(1)- tx_add_input ---->|                  |                  MsgSentChange                   +
+//     |            |<-(2)- tx_complete ------|                  |                ReceivedTxComplete                +
+//     |            |--(3)- tx_add_output --->|                  |                  MsgSentChange                   +
+//     |            |<-(4)- tx_complete ------|                  |                ReceivedTxComplete                +
+//     |            |--(5)- tx_add_input ---->|                  |                  MsgSentChange                   +
+//     |   Holder   |<-(6)- tx_add_input -----|   Counterparty   |                MsgReceivedChange                 +
+//     |            |--(7)- tx_remove_output >|                  |                  MsgSentChange                   +
+//     |            |<-(8)- tx_add_output ----|                  |                MsgReceivedChange                 +
+//     |            |--(9)- tx_complete ----->|                  |                  SentTxComplete                  +
+//     |            |<-(10) tx_complete ------|                  |                NegotiationComplete               +
+//     +------------+                         +------------------+--------------------------------------------------+
+
+/// Negotiation states that can send & receive `tx_(add|remove)_(input|output)` and `tx_complete`
+trait State {}
+
+/// Category of states where we have sent some message to the counterparty, and we are waiting for
+/// a response.
+trait MsgSentState: State {
+	fn into_negotiation_context(self) -> NegotiationContext;
+}
+
+/// Category of states that our counterparty has put us in after we receive a message from them.
+trait MsgReceivedState: State {
+	fn into_negotiation_context(self) -> NegotiationContext;
+}
+
+// This macro is a helper for implementing the above state traits for various states subsequently
+// defined below the macro.
+macro_rules! define_state {
+	(MSG_SENT_STATE, $state: ident, $doc: expr) => {
+		define_state!($state, NegotiationContext, $doc);
+		impl MsgSentState for $state {
+			fn into_negotiation_context(self) -> NegotiationContext {
+				self.0
+			}
+		}
+	};
+	(MSG_RECEIVED_STATE, $state: ident, $doc: expr) => {
+		define_state!($state, NegotiationContext, $doc);
+		impl MsgReceivedState for $state {
+			fn into_negotiation_context(self) -> NegotiationContext {
+				self.0
+			}
+		}
+	};
+	($state: ident, $inner: ident, $doc: expr) => {
+		#[doc = $doc]
+		#[derive(Debug)]
+		struct $state($inner);
+		impl State for $state {}
+	};
+}
+
+define_state!(
+	MSG_SENT_STATE,
+	MsgSentChange,
```

Comment:
Naming would be more consistent with `SentTxComplete` and a bit more readable IMO if this were named `SentChangeMsg`. Likewise for others.